### ISSUE TITLE
"for arrays in uproot.iterate(...)" raises value exception 

### ISSUE THREAD

**Issue by rklasen** [scikit-hep/uproot#225]
Hi,

first off, thank you for this great tool.

I'm trying to iterate over multiple root files using this example: https://github.com/scikit-hep/uproot/blob/master/README.rst#iterating-over-files-like-tchain, and I'm using this following code snippet for it:

```python
def sortAll():
    targetDir = './numpyPairs/'
    if not os.path.exists(targetDir):
        os.makedirs(targetDir)
    print('using uproot.iterate():')
    try:
        # open the root trees in a TChain-like manner
        for arrays in uproot.iterate('./RootPairs/Lumi_Pairs*.root', 'pndsim', [b'PndLmdHitPair._overlapID', b'PndLmdHitPair._hit1', b'PndLmdHitPair._hit2'], entrysteps=1000000):
            print('processing file...')
            sortPairs(arrays)
    except Exception as e:
        print('error occured:')
        print(e)
    
    print('using naive iteration:')
    # use caveman tactics
    files = [
        './RootPairs/Lumi_Pairs_1000000.root',
        './RootPairs/Lumi_Pairs_0-copy.root'
    ]
    for file in files:
        print('processing file...')
        arrays = uproot.open(file)['pndsim'].arrays([b'PndLmdHitPair._overlapID', b'PndLmdHitPair._hit1', b'PndLmdHitPair._hit2'])
        sortPairs(arrays)
    print('done')
```

I'm using two separate ways to loop over the two root files, because the `uproot.iterate()` function raises a value exception, whereas the naive version works as expected:

```
branch b'PndLmdHitPair._hit2' interpreted as asjagged(asobj(<uproot_methods.classes.TVector3.Methods>)) in './RootPairs/Lumi_Pairs_1000000.root', but as asjagged(asobj(<uproot_methods.classes.TVector3.Methods>)) in './RootPairs/Lumi_Pairs_0-copy.root'
```

However, from what I can see, the `asjagged(asobj(<uproot_methods.classes.TVector3.Methods>))` objects are identical in both cases, so `tree.py` should not raise this exception. To confirm the problem is not in my data, I symlinked one of the files to the other, so in both cases the content of the root file is identical:

```
$ ls -lh RootPairs/
total 86M
drwxr-xr-x 2 roman roman 4,0K Feb  8 15:34 exclude
lrwxrwxrwx 1 roman roman   23 Feb  8 15:34 Lumi_Pairs_0-copy.root -> Lumi_Pairs_1000000.root
-rw------- 1 roman roman  86M Feb  8 12:57 Lumi_Pairs_1000000.root
```

But the problem persists. When I iterate over the files like in the for loop, no error occures.

Best wishes,
Roman Klasen

Edit: I'm using the uproot version from pip3:

```
$ python3
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import uproot
>>> print(uproot.__version__)
3.4.4
>>> 
```

**Comment by jpivarski** [comment]
What's happening here is that the branch structure in the second file is different from the branch structure in the first file, though they can both be interpreted as `TVector3`. This class has been stable for some time, so I can only imagine that it's some kind of trivial change we don't care about.

(Note: copying a file can change the branch structure because the ROOT version used for copying may be different from the ROOT version used to generate the original file, and thus they get written with different versions of the C++ classes.)

You should be able to see that they are incompatible with

```python
tree1 = uproot.open('./RootPairs/Lumi_Pairs_1000000.root')['pndsim']
tree2 = uproot.open('./RootPairs/Lumi_Pairs_0-copy.root')['pndsim']
interp1 = tree1['PndLmdHitPair._hit2'].interpretation
interp2 = tree2['PndLmdHitPair._hit2'].interpretation
interp1.compatible(interp2)     # should be False
```

To figure out why, print `interp1.content`, `interp1.content.content` and `interp2.content`, `interp2.content.content`. I'm curious to see what's actually different between these two classes.

Meanwhile, the right thing to do is probably to widen the definition of "compatible" for `asobj` and `asgenobj`. For these types, you're accessing them through high-level Python properties and methods that probably don't depend on whatever's different from one file to the next. I'll start on that, but I really am curious to know what changed (and it might be surprising; it might change my course of action).

**Comment by jpivarski** [comment]
I started PR #226 with a potential solution to this problem, but I'm not going to merge it until I understand better what's going on in your files.

The Right Thing To Do is probably to consider `asobj` and `asgenobj` branches compatible if they are represented by a class with the same name (e.g. `TVector3` versus `TVector3`), and that's what this PR would do.

**Comment by rklasen** [comment]
Hi, thanks for your quick reply. I wanted to make sure the branch structure in both files is identical, so I didn't copy the file, but instead created a symlink to it:

```
$ ln -s Lumi_Pairs_1000000.root Lumi_Pairs_1000000-symlink.root 
$ ls -lah
total 87780
drwxr-xr-x 3 roman roman     4096 Feb  9 10:34 ./
drwxr-xr-x 5 roman roman     4096 Feb  8 14:29 ../
-rw------- 1 roman roman 89872428 Feb  8 12:57 Lumi_Pairs_1000000.root
lrwxrwxrwx 1 roman roman       23 Feb  9 10:34 Lumi_Pairs_1000000-symlink.root -> Lumi_Pairs_1000000.root
```

Now, I check for compatibility:

```
def showCompatibility():
    print('showing interpretation of both files:')
    tree1 = uproot.open('./RootPairs/Lumi_Pairs_1000000.root')['pndsim']
    tree2 = uproot.open('./RootPairs/Lumi_Pairs_1000000-symlink.root')['pndsim']
    interp1 = tree1['PndLmdHitPair._hit2'].interpretation
    interp2 = tree2['PndLmdHitPair._hit2'].interpretation
    print(interp1.content)
    print(interp2.content)
    print(interp1.content.content)
    print(interp2.content.content)
    print(interp1.compatible(interp2))     # should be False
```

And the result is false, just you you said:

```
$ ./readRootPairFiles.py 
showing interpretation of both files:
asobj(<uproot_methods.classes.TVector3.Methods>)
asobj(<uproot_methods.classes.TVector3.Methods>)
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
False
```

However, I really don't understand why, since this time, there really is only one file. I uploaded it to a Dropbox share so you can inspect the file, maybe I'm doing something unexpected during file creation: [Dropbox Link | LumiPairs](https://www.dropbox.com/s/y125gqejrtlkoko/Lumi_Pairs_1000000.root?dl=0). I also uploaded the entire python script [here](https://www.dropbox.com/s/g0czjjd8eaxfrxk/readRootPairFiles.py?dl=0), in case I made a mistake there.

Thanks!

**Comment by jpivarski** [comment]
This is super-weird. For `interp1.compatible(interp2)` to be false, I expected them to differ internally, since that's what the old (3.4.4, without PR #226) implementation did. But that was also based on the supposition that the two files were _copies_, not _simlinks_. (I didn't read your original post carefully enough.) If they're simlinked, they're identical. Your print-out shows that the internal structure is identical. There's no reason they could be flagged as incompatible.

So I downloaded your file and made a simlink of it. I find that `interp1.compatible(interp2)` is **true**, even when I revert the implementation to 3.4.4.

Something's very fishy. I looked at your script: it's exactly what I'm doing.

I can't reproduce it, and I can see no reason how you could be getting that error, both from the theoretical standpoint (symlinked files really are identical! 'cause there's not even two of them!) and from the print-outs of the inner content of that interpretation, which is what the old interpretation used to determine incompatibility.

So... Is it considered compatible when you use the version in the PR? That's not addressing the fundamental weirdness, but it's not a bad idea to loosen the definition of compatibility when you're accessing it through a high-level class like TVector3, anyway.

Are you using Windows? Maybe simlinks are different in Windows?

**Comment by rklasen** [comment]
No, I'm using Linux. I tested this on two different real machines and one virtual machine, the compatible() call always returns False: 

**machine at work**
$ cat /etc/issue
Ubuntu 18.04.1 LTS \n \l

$ uname -a
Linux roklasen-dt 4.15.0-43-generic #46-Ubuntu SMP Thu Dec 6 14:45:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**my machine at home**

$ cat /etc/issue
Ubuntu 18.04.1 LTS \n \l

$ uname -a
Linux z170-ubuntu 4.18.0-15-generic #16-Ubuntu SMP Thu Feb 7 10:56:39 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

** brand new Debian VM**

$ uname -a
Linux debian 4.9.0-8-amd64 #1 Debian 4.9.130.-2 (2018-10-27) x86_64 GNU/Linux
$ cat /etc/issue
Debian GNU/Linux 9 \n \l


The output is the same on *every* machine:

```
showing interpretation of both files:
asobj(<uproot_methods.classes.TVector3.Methods>)
asobj(<uproot_methods.classes.TVector3.Methods>)
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
False
```

I realize that you can't really debug a problem that doesn't occur on your machine, and it's more than likely that my test machines have differntly python installations and other packages installed, so I created this Dockerfile. It uses Alpine Linux, installs uproot and uses my minimal example from above.

```
FROM alpine:latest

RUN apk add curl bash											&& \
    apk add --update --no-cache python3 py-pip py-numpy		&& \
    find / -type d -name __pycache__ -exec rm -r {} +   	&& \
    rm -r /usr/lib/python*/ensurepip                    	&& \
    rm -r /usr/lib/python*/lib2to3                     		&& \
    rm -r /usr/lib/python*/turtledemo                   	&& \
    rm /usr/lib/python*/turtle.py                      		&& \
    rm /usr/lib/python*/webbrowser.py                   	&& \
    rm /usr/lib/python*/doctest.py                      	&& \
    rm /usr/lib/python*/pydoc.py                        	&& \
    rm -rf /root/.cache /var/cache /usr/share/terminfo
RUN pip3 install uproot
RUN mkdir uproot																	&& \
    cd uproot																		&& \
    curl -L -O https://www.dropbox.com/s/g0czjjd8eaxfrxk/readRootPairFiles.py		&& \
    mkdir RootPairs																	&& \
    cd RootPairs																	&& \
    curl -L -O https://www.dropbox.com/s/y125gqejrtlkoko/Lumi_Pairs_1000000.root	&& \
    ln -s Lumi_Pairs_1000000.root Lumi_Pairs_1000000-symlink.root					&& \
    cd ../																			&& \
    chmod +x readRootPairFiles.py
```

It uses the same dropbox links from above, no changes. Build the docker container with: `sudo docker build -t test/uproot .` The output is quite long, but it's only build info.

Then run the container and my script in it: `sudo docker run test/uproot /bin/sh -c 'cd /uproot; ./readRootPairFiles.py'` It should say:

```
$ sudo docker run test/uproot /bin/sh -c 'cd /uproot; ./readRootPairFiles.py'
showing interpretation of both files:
asobj(<uproot_methods.classes.TVector3.Methods>)
asobj(<uproot_methods.classes.TVector3.Methods>)
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
False
```

I couldn't test the new version of uproot yet.

**Comment by rklasen** [comment]
Hi, your last PR fixed the issue. I don't know if it makes sense to look exactly why the `compatible()` call failed before, but since the issue is fixed now, I'm closing it. Thanks!

**Comment by jpivarski** [comment]
I'm glad that this fixed it, but to be honest, it didn't really address the underlying problem. However, I can't reproduce the underlying problem, so I'm at a bit of a loss.

**Comment by rklasen** [comment]
Did the Dockerfile work for you? Since the uproot version in pip is now 3.4.5, we only need to change the line `pip install [...]` to `RUN pip3 install uproot==3.4.4` to reliably build a docker image that reproduces the error. I just tested it on my docker machine. I'd be glad to help anyway I can, by providing verbose debug output for example.

**Comment by jpivarski** [comment]
Oh, I've never run docker. (Not locally; it's used in many services, I know.) I probably wouldn't be able to on this Chromebook, though.

**Comment by rklasen** [comment]
You're right, docker on a chromebook is probably not a good idea :sweat_smile: . You can use [play with docker](https://labs.play-with-docker.com) for that, it's practically made fur just this scenario. Simply create an account and you can add instances (virtual machines which have docker installed). You'll get a standard linux terminal, only inside the web browser. You can use vim to create the `Dockerfile` (it has to have this exact name), then build and run the container:

```
[node1] (local) root@192.168.0.58 ~
$ vim Dockerfile
[node1] (local) root@192.168.0.58 ~
$ cat Dockerfile
FROM alpine:latest

RUN apk add curl bash                     && \
    apk add --update --no-cache python3 py-pip py-numpy   && \
        find / -type d -name __pycache__ -exec rm -r {} +     && \
            rm -r /usr/lib/python*/ensurepip                      && \
                rm -r /usr/lib/python*/lib2to3                        && \
                    rm -r /usr/lib/python*/turtledemo                     && \
                        rm /usr/lib/python*/turtle.py                         && \
                            rm /usr/lib/python*/webbrowser.py                     && \
                                rm /usr/lib/python*/doctest.py                        && \
                                    rm /usr/lib/python*/pydoc.py                          && \
                                        rm -rf /root/.cache /var/cache /usr/share/terminfo
                                        RUN pip3 install uproot
                                        RUN mkdir uproot                                  && \
                                            cd uproot                                   && \
                                                curl -L -O https://www.dropbox.com/s/g0czjjd8eaxfrxk/readRootPairFiles.py   && \
                                                    mkdir RootPairs                                 && \
                                                        cd RootPairs                                  && \
                                                            curl -L -O https://www.dropbox.com/s/y125gqejrtlkoko/Lumi_Pairs_1000000.root  && \
                                                                ln -s Lumi_Pairs_1000000.root Lumi_Pairs_1000000-symlink.root         && \
                                                                    cd ../                                      && \
                                                                        chmod +x readRootPairFiles.py
[node1] (local) root@192.168.0.58 ~
```

(the formatting is kind of weird here, but I think it's clear what it does). Build the docker image:

```
$ docker build -t test/uproot .
Sending build context to Docker daemon  1.739MB
Step 1/4 : FROM alpine:latest
[...]
Successfully tagged test/uproot:latest
```

And run the container with my script:

```
$ docker run test/uproot /bin/sh -c 'cd /uproot; ./readRootPairFiles.py'
showing interpretation of both files:
asobj(<uproot_methods.classes.TVector3.Methods>)
asobj(<uproot_methods.classes.TVector3.Methods>)
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
astable(asdtype("[(' fBits', '>u8'), (' fUniqueID', '>u8'), ('fX', '>f8'), ('fY', '>f8'), ('fZ', '>f8')]", "[('fX', '<f8'), ('fY', '<f8'), ('fZ', '<f8')]"))
True
```

This is with `RUN pip install uproot` without version number, that's why the evaluation is `True`. Simply change that line to specify `uproot==3.4.4`. I just tested it on Play with Docker, works really nice. The only caveat is that your instances are destroyed after 4 hours.

**Comment by jpivarski** [comment]
Thanks! I was just writing to ask how to do that.

**Comment by jpivarski** [comment]
Found it! The new version actually does fix the fundamental issue, though it did so accidentally. The old code was

```python
    def compatible(self, other):
        return self.content.compatible(other)
```

instead of

```python
    def compatible(self, other):
        return self.content.compatible(other.content)
```

and no matter how many times I looked at that while debugging before, I didn't see it. (When I was typing it on the commandline and when I "reverted" it to the old version, I entered the correct code, not the old code.) The new code _accidentally_ fixes the issue because it completely replaces that line with a different intention.

Thanks for your help with Docker! (Learning new things is a good thing.  :)
